src/core/kvs.jl

```diff
@@ -0,0 +1,94 @@
+mutable struct Transaction
+    handle::Ptr{API.flux_kvs_txn_t}
+    function Transaction()
+        handle = API.flux_kvs_txn_create()
+        this = new(handle)
+        finalizer(this) do txn
+            API.flux_kvs_txn_destroy(txn.handle)
+        end
+        return this
+    end
+end
+Base.unsafe_convert(::Type{Ptr{API.flux_kvs_txn_t}}, txn::Transaction) = txn.handle
+
+function commit(flux::Flux, txn::Transaction, ns=C_NULL; flags=0)
+    handle = API.flux_kvs_commit(flux, ns, flags, txn)
+    Libc.systemerror("flux_kvs_commit", handle == C_NULL)
+    fut = Future(handle)
+    fut.refs[txn] = nothing # root txn in fut
+    return fut
+end
+
+function fence(flux::Flux, txn::Transaction, name, nprocs, ns=C_NULL; flags=0)
+    handle = API.flux_kvs_fence(flux, ns, flags, name, nprocs, txn)
+    Libc.systemerror("flux_kvs_fence", handle == C_NULL)
+    fut = Future(handle)
+    fut.refs[txn] = nothing # root txn in fut
+    return fut
+end
+
+mutable struct KVS
+    flux::Flux
+    current_txn::Transaction
+    lock::Base.ReentrantLock
+    function KVS(flux::Flux)
+        new(flux, Transaction(), Base.ReentrantLock())
+    end
+end
+function exchange_transaction!(kvs::KVS)
```

Comment:
I'm guessing this is modeled after our python bindings?  The way it is done here with a global txn is totally fine. I know that's how our py bindings do it currently.  I think our python bindings for the KVS are probably due for an update; they were originally written before transactions are a thing. Going forward, I think we want to add context manager support to python's transactions so that you can do things like:
```python
with flux.kvs.transaction() as t:
    t.put("foo", "bar")
    t.mkdir("baz")
```

That way you don't have to worry about commit'ing the transaction and there is no global transaction to worry about.  Not sure if something like that maps well to Julia, but I thought I'd mention it just in case.

Reply:
Ah I didn't look at the Python bindings for this. Convergent evolution ;)